text/0000-wasm-smart-contracts.md

```diff
@@ -0,0 +1,770 @@
+improve over time as Wasm popularity grows, or it may persist into the future.
+
+The use of named contracts is substantially different than other smart contract
+systems such as Ethereum, which instead use a hash of the contract as the
+unique identifier. The design was chosen because it provides a friendlier
+method of versioning smart contracts for permissioned networks. However, it
+would not be a suitable approach for a public non-permissioned network; thus,
+supporting such networks will require substantial design changes in the future.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The design presented is intended to maximize compatibility with the existing
+Sawtooth transaction processor approach. It is intended to extend the full
+capability of transaction processors into the realm of deployable on-chain
+smart contracts. To achieve this, it was clear that a permissioning system
+would be required to constrain deployments.
+
+The only existing on-chain smart contract system for Sawtooth is Seth, which
+implements Ethereum compatibility via a JSON-RPC interface and execution of
+smart contracts in the EVM. Seth does not currently have access to all of the
+features present in Sawtooth transaction processors, nor can Seth smart
+contracts access the entire global state. Is is therefore impossible today, for
+example, to write a Seth contract which interacts with the Intkey namespace.
+It would be possible to extend Seth with a set of built-in smart contracts
+which provided native access to Sawtooth features and/or non-Seth global state.
+However, the inherent complexity, both from an implementation and a user
+experience perspective, makes this solution less desirable. It would also
+provide no natural transition path from Sawtooth transaction processors.
+
+# Prior art
+[prior-art]: #prior-art
+
+Sawtooth currently relies on Transaction Processors that run in a separate
+process. The Sabre API for the smart contracts will closely match that of the
+current transaction processor API.
+
+Sawtooth Sabre follows the intent of Sawtooth [Seth][seth], being able to run
+smart contracts on Sawtooth, but removes the dependencies on an Ethereum
+Virtual Machine.
+
+This design borrows some ideas from [Ethereum][ethereum] smart contracts run on
+the Ethereum Virtual Machine as well as [Parity][parity] and
+[Parity Wasm][wasmi].
+
+[seth]: https://github.com/hyperledger/sawtooth-seth
+[ethereum]:https://www.ethereum.org/greeter
+[parity]:https://github.com/paritytech/parity/tree/27c32d3629e4b70b63ee37736147318035f48792/ethcore/wasm
+[wasmi]:https://github.com/paritytech/wasmi
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+The current design does not describe a method for limiting execution (in
+a manner similar to gas in Ethereum). It is therefore not yet suitable for
+deployment of code by potentially malicious actors. This will be addressed in
+a future RFC.
```

Comment:
Runaway code isn't restricted to just malicious actors. It can be the result of inadvertent programming errors. Even if it's not initially implemented, I think there either needs to be a premise how to do the execution accounting or a means of stopping a runaway contract.

Reply:
We don't handle this with transaction processors, so even without this feature designed and implemented, we are on-par with current state.

One possible way to implement this is to modify the bytecode and add the equivalent of a use_gas() call at the right points. Another would be to abort the interpreter purely based on the number of instructions it has executed. Currently, we have not taken a position on which path would be better.

Reply:
I think the difference is that system admins need to take an active role in launching new transaction processors, and it's possible to gate adding TPs through the permissioning system. The concept of operations is that the consortium would review and ratify code before deploying it on the network.
Sabre would alleviate the logistical aspects of launching apps on the network but should offer some controls like the TP process to mitigate some risks. I added a question about permissioning earlier in the doc just now.
Permissioning might still be a bit less protection than status quo because a permissioned actor can still launch unreviewed code.
If the long term solution is instruction accounting, I would appreciate some dialog on the feasibility within wasm or some way to judge if we really have line of sight.